added socket_timeout to HTTPConnection

johnwheeler opened the pull request:
Hello Cory,

This is my first attempt. Please let me know your thoughts. 

Thank you

Lukasa commented on hyper/http11/connection.py:
```diff
@@ -117,6 +122,7 @@ def connect(self):
                 port = self.proxy_port
 
             sock = socket.create_connection((host, port), 5)
```
We probably want to use the timeout in `create_connection` too: see [the docs](https://docs.python.org/2/library/socket.html#socket.create_connection) for how to do it.

Lukasa commented on hyper/http20/connection.py:
```diff
@@ -361,6 +366,7 @@ def connect(self):
                 port = self.proxy_port
 
             sock = socket.create_connection((host, port))
```
Same note as above.

Lukasa commented on test/test_abstraction.py:
```diff
@@ -10,22 +10,23 @@ def test_h1_kwargs(self):
             'test', 443, secure=False, window_manager=True, enable_push=True,
             ssl_context=True, proxy_host=False, proxy_port=False,
-            other_kwarg=True
+            socket_timeout=None, other_kwarg=True
```
Ideally we want to set this to a non-None value to confirm it's not getting defaulted. Let's try....`10`.

Lukasa commented on test/test_http11.py:
```diff
@@ -511,6 +512,13 @@ def read(self, size):
                'Got: {}'.format(int) in str(exc_info)
 
+    def test_exception_raised_on_socket_timeout(self):
+        c = HTTP11Connection('httpbin.org', socket_timeout=0.01)
```
Generally it's good to set this value comedically small so that running on faster machines doesn't accidentally get this to work.

johnwheeler replied:
Thank you for taking the time to make these comments. I will adjust the changes above. For this line comment, the URL I am requesting is designed to block for n seconds

 c.request('GET', '/delay/1')

From the docs on httpbin "Delays responding for min(n, 10) seconds."

so I think that would save us from even Moore's Law. Please advise if I'm overlooking something.

Lukasa replied:
Ah, yes, sorry, you're right.

Lukasa commented:
Thanks for this! In the first instance I've provided a few specific notes in the diff. Please take a look at those and see if you can apply them. Let me know if you need anything extra!

johnwheeler commented:
I made the changes. I removed the test case I added because since `socket_timeout` is used for both read and connect timeout, the read timeout can't be exercised (no read if no connect). this can be solved by having separate timeout values. If that's your preference, please let me know. I am indifferent at this time.

johnwheeler commented:
Ugh - I'm sorry for being wishy-washy. I was looking at how requests does it

http://docs.python-requests.org/en/master/user/advanced/ Timeout section

Maybe we should have separate read and connect timeouts with this tuple based approach? 

I'm in no hurry to get this merged and want to do it the right way, so please take your time to think about this, and I'll do the work.

Lukasa commented:
I'm open to having separate timeout values, sure. =)
